x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java

```diff
@@ -180,6 +156,33 @@ private Object unwrapMultiValue(Object values) {
         throw new SqlIllegalArgumentException("Type {} (returned by [{}]) is not supported", values.getClass().getSimpleName(), fieldName);
     }
 
+    private Object extractGeoShape(Object values) {
+        Object value;
+        if (values instanceof List) {
+            if (((List<?>) values).size() > 0) {
```

Comment:
I don't think this is the correct approach here. If I'm not mistaken you are extracting here the first value from a field that has an array of geo shapes. SQL has added recently a "multi field leniency" attribute to both REST and drivers support: https://www.elastic.co/guide/en/elasticsearch/reference/6.7/sql-rest.html#sql-rest-fields.
Code-wise, this leniency is handled [here](https://github.com/elastic/elasticsearch/blob/dd3e12275c70444e13461fe53da7c565e5f3db3e/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java#L210) and [here](https://github.com/elastic/elasticsearch/blob/dd3e12275c70444e13461fe53da7c565e5f3db3e/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/FieldHitExtractor.java#L138) at extraction time. I think you need to include `extractGeoShape()` in the array leniency logic probably only in `unwrapMultiValue` method.

Also, would be good to see some tests with arrays of geo shapes and not only in one-level fields, but also multi nested fields (not the `nested` field type): `field.subfield1.subfield2.my_geo_shapes_array`.

Reply:
Sorry for the delay and thanks for bringing this to my attention. That's very good point and it turned out to be quite tricky to get right. I am still working on it. I might need to make some changes in master first thought to change the way extractSource handles `a.b.c.` names. I hope to open this pull request monday.